File: client/jetbrains/src/main/java/com/sourcegraph/cody/chat/MessagePanel.java
package com.sourcegraph.cody.chat;

import com.intellij.openapi.ui.VerticalFlowLayout;
import com.intellij.ui.ColorUtil;
import com.intellij.util.ui.UIUtil;
import com.sourcegraph.cody.api.Speaker;
import com.sourcegraph.cody.ui.Colors;
import java.awt.*;
import javax.swing.*;
import javax.swing.border.Border;
import org.jetbrains.annotations.NotNull;

public class MessagePanel extends JPanel {

  private final boolean isHuman;
  private final int gradientWidth;

  public MessagePanel(Speaker speaker, int gradientWidth) {
    super();
    this.gradientWidth = gradientWidth;
    Border emptyBorder = BorderFactory.createEmptyBorder(0, 0, 0, 0);
    Color background = UIUtil.getPanelBackground();
    Border topBorder =
        BorderFactory.createMatteBorder(1, 0, 0, 0, ColorUtil.brighter(background, 2));
    Border bottomBorder =
        BorderFactory.createMatteBorder(0, 0, 1, 0, ColorUtil.brighter(background, 3));
    Border topAndBottomBorder = BorderFactory.createCompoundBorder(topBorder, bottomBorder);

    isHuman = speaker.equals(Speaker.HUMAN);
    this.setBorder(isHuman ? emptyBorder : topAndBottomBorder);
    this.setBackground(isHuman ? ColorUtil.darker(background, 2) : background);
    this.setLayout(new VerticalFlowLayout(VerticalFlowLayout.TOP, 0, 0, true, false));
  }

  @Override
  protected void paintComponent(@NotNull Graphics g) {
    super.paintComponent(g);
    paintLeftBorderGradient(g);
  }

  private void paintLeftBorderGradient(Graphics g) {
    if (isHuman) return;
    int panelHeight = getHeight();
    int halfOfHeight = panelHeight / 2;

    GradientPaint firstPartGradient =
        new GradientPaint(0, 0, Colors.PURPLE, 0, halfOfHeight, Colors.ORANGE);
    GradientPaint secondPartGradient =
        new GradientPaint(0, halfOfHeight, Colors.ORANGE, 0, panelHeight, Colors.CYAN);
    final Graphics2D g2d = (Graphics2D) g;
    g2d.setPaint(firstPartGradient);
    g2d.fillRect(0, 0, gradientWidth, halfOfHeight);
    g2d.setPaint(secondPartGradient);
    g2d.fillRect(0, halfOfHeight, gradientWidth, panelHeight);
  }
}

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/chat/MessagePanel.kt
package com.sourcegraph.cody.chat

import com.intellij.openapi.project.Project
import com.intellij.ui.ColorUtil
import com.intellij.util.ui.SwingHelper
import com.intellij.util.ui.UIUtil
import com.sourcegraph.cody.api.Speaker
import com.sourcegraph.cody.ui.HtmlViewer.createHtmlViewer
import java.awt.Color
import javax.swing.JEditorPane
import javax.swing.JPanel
import org.commonmark.ext.gfm.tables.TablesExtension
import org.commonmark.node.Node
import org.commonmark.parser.Parser
import org.commonmark.renderer.html.HtmlRenderer

class MessagePanel(
    private val chatMessage: ChatMessage,
    private val project: Project,
    private val parentPanel: JPanel,
    private val gradientWidth: Int,
) : PanelWithGradientBorder(gradientWidth, chatMessage.speaker) {
  private var lastMessagePart: MessagePart? = null

  init {
    val markdownNodes: Node = markdownParser.parse(chatMessage.displayText)
    markdownNodes.accept(MessageContentCreatorFromMarkdownNodes(this, htmlRenderer))
  }

  fun updateContentWith(message: ChatMessage) {
    val markdownNodes = markdownParser.parse(message.displayText)
    val lastMarkdownNode = markdownNodes.lastChild
    if (lastMarkdownNode.isCodeBlock()) {
      val (code, language) = lastMarkdownNode.extractCodeAndLanguage()
      addOrUpdateCode(code, language)
    } else {
      val nodesAfterLastCodeBlock = markdownNodes.findNodeAfterLastCodeBlock()
      val renderedHtml = htmlRenderer.render(nodesAfterLastCodeBlock)
      addOrUpdateText(renderedHtml)
    }
  }

  fun addOrUpdateCode(code: String, language: String?) {
    val lastPart = lastMessagePart
    if (lastPart is CodeEditorPart) {
      lastPart.updateCode(project, code, language)
    } else {
      addAsNewCodeComponent(code, language)
    }
  }

  private fun addAsNewCodeComponent(code: String, info: String?) {
    val codeEditorComponent =
        CodeEditorFactory(project, parentPanel, gradientWidth).createCodeEditor(code, info)
    this.lastMessagePart = codeEditorComponent
    add(codeEditorComponent.component)
  }

  fun addOrUpdateText(text: String) {
    val lastPart = lastMessagePart
    if (lastPart is TextPart) {
      lastPart.updateText(text)
    } else {
      addAsNewTextComponent(text)
    }
  }

  private fun addAsNewTextComponent(renderedHtml: String) {
    val textPane: JEditorPane = createHtmlViewer(getInlineCodeBackgroundColor(chatMessage.speaker))
    SwingHelper.setHtml(textPane, renderedHtml, UIUtil.getLabelForeground())
    val textEditorComponent = TextPart(textPane)
    this.lastMessagePart = textEditorComponent
    add(textEditorComponent.component)
  }

  private fun getInlineCodeBackgroundColor(speaker: Speaker): Color {
    return if (speaker == Speaker.ASSISTANT) ColorUtil.darker(UIUtil.getPanelBackground(), 3)
    else ColorUtil.brighter(UIUtil.getPanelBackground(), 3)
  }

  companion object {
    private val extensions = listOf(TablesExtension.create())

    private val markdownParser = Parser.builder().extensions(extensions).build()
    private val htmlRenderer =
        HtmlRenderer.builder().softbreak("<br />").extensions(extensions).build()
  }
}
